proc-macros/src/lib.rs

```diff
@@ -111,212 +110,278 @@ pub fn rpc_client_api(input_token_stream: TokenStream) -> TokenStream {
+///     async fn sub(&self, subscription_sink: SubscriptionSink) -> Result<Subscription<String>, Error> {
+///         // ...
+///     }
+/// }
+///
+/// impl<T> RpcClient for T where T: SubscriptionClient {}
+/// ```
+///
+/// ## Attributes
+///
+/// ### `rpc` attribute
+///
+/// `rpc` attribute is applied to a trait in order to turn it into an RPC implementation.
+///
+/// **Arguments:**
+///
+/// - `server`: generate `<Trait>Server` trait for the server implementation.
+/// - `client`: generate `<Trait>Client` extension trait that makes RPC clients to invoke a concrete RPC
+///   implementation methods conveniently.
+/// - `namespace`: add a prefix to all the methods and subscriptions in this RPC. For example, with namespace
+///   `foo` and method `spam`, the resulting method name will be `foo_spam`.
+///
+/// **Trait requirements:**
+///
+/// Trait wrapped with an `rfc` attribute **must not**:
+///
+/// - have associated types or constants;
+/// - have Rust methods not marked with either `method` or `subscription` attribute;
+/// - be empty.
+///
+/// At least one of the `server` or `client` flags must be provided, otherwise the compilation will err.
+///
+/// ### `method` attribute
+///
+/// `method` attribute is used to define an RPC method.
+///
+/// **Arguments:**
+///
+/// - `name` (mandatory): name of the RPC method. Does not have to be the same as the Rust method name.
+///
+/// **Method requirements:**
+///
+/// Rust method marked with `method` attribute, **may**:
+///
+/// - be either `async` or not;
+/// - have input parameters or not;
+/// - have return value or not (in the latter case, it will be considered a notification method).
+///
+/// ### `subscription` attribute
+///
+/// **Arguments:**
+///
+/// - `name` (mandatory): name of the RPC method. Does not have to be the same as the Rust method name.
+/// - `unsub` (mandatory): name of the RPC method to unsubscribe from the subscription. Must not be the same as `name`.
+/// - `item` (mandatory): type of items yielded by the subscription. Note that it must be the type, not string.
+///
+/// **Method requirements:**
+///
+/// Rust method marked with `subscription` attribute **must**:
+///
+/// - be synchronous;
+/// - not have return value.
+///
+/// Rust method marked with `subscription` attribute **may**:
+///
+/// - have input parameters or not.
+///
+/// ## Full workflow example
+///
+/// ```rust
+/// //! Example of using proc macro to generate working client and server.
+///
+/// use std::net::SocketAddr;
+///
+/// use futures_channel::oneshot;
+/// use jsonrpsee::{ws_client::*, ws_server::WsServerBuilder};
+///
+/// // RPC is moved into a separate module to clearly show names of generated entities.
+/// mod rpc_impl {
+///     use jsonrpsee::{proc_macros::rpc, types::async_trait, ws_server::SubscriptionSink};
+///
+///     // Generate both server and client implementations, prepend all the methods with `foo_` prefix.
+///     #[rpc(client, server, namespace = "foo")]
```

Comment:
I like the namespace attribute, Is it possible to add any number of traits for the server to implement?

Reply:
> Is it possible to add any number of traits for the server to implement?

Can you elaborate on what do you mean?

Reply:
Ah, do you mean whether one `struct` can implement several server traits (like `#[rpc(server)] trait Foo` and `#[rpc(server) trait Bar`)? If so, it sure can.

Reply:
exactly.

so I suppose it's not possible to conflicting `functions names` in different traits then?

For example:

```rust
#[rpc(client, server, namespace = "foo")]
pub trait RpcNamespaceFoo {
	#[method(name = "foo")]
	async fn async_method(&self, param_a: u8, param_b: String) -> u16;
}

#[rpc(client, server, namespace = "bar")]
pub trait RpcNamespaceBar {
	#[method(name = "foo")]
	async fn async_method(&self, param_a: u8, param_b: String, x: usize) -> u16;
}
```

Reply:
As long as prefixes are different, I don't think there should be a problem. When you will call `RpcNamespaceFoo::into_rpc(self)`, it will register method `foo_foo`, and when you will call `RpcNamespaceBar::into_rpc(self)`, it will register method `bar_foo`, so there will be no name conflict.

Reply:
The same applies to the client. You will be able to call `RpcNamespaceFooClient::async_method(&client, 10, "a".into())` and `RpcNamespaceBarClient::async_method(&client, 10,  "a".into(), 20)`, and it will call `foo_foo` and `bar_foo` correspondingly.

Reply:
all right, you are right I just tried and it works.

FWIW, the `server` in this context is `RpcServerImpl` or the type that implements the traits or one of them. I got that wrong myself